### PR TITLE
SI-6035 - retain specialized flag in separate compilation.

### PR DESCRIPTION
**Pull request by axel22**
This solution does not pickle the `specialize` flag on the symbol of the
specialized type parameter.
Instead, it recreates the `specialize` flag on a type parameter during the
unpickling process if it sees the specialize annotation on that type parameter.

Now with separate compilation tests.

Review by @VladUreche, @paulp, @gkossakowski.

### CODE REVIEW

**Comment by VladUreche** [comment]
Awesome! LGTM

File: src/compiler/scala/tools/nsc/transform/UnCurry.scala [hunk 1]
```diff
@@ -566,10 +566,10 @@ abstract class UnCurry extends InfoTransform
       val sym = tree.symbol
       // Take a pass looking for @specialize annotations and set all
       // their SPECIALIZE flags for cheaper recognition.
-      if ((sym ne null) && (sym.isClass || sym.isMethod)) {
-        for (tp <- sym.typeParams ; if tp hasAnnotation SpecializedClass)
-          tp setFlag SPECIALIZED
-      }
+      // if ((sym ne null) && (sym.isClass || sym.isMethod)) {
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/UnCurry.scala]: Why not remove it completely instead of commenting out?

**Comment by gkossakowski** [comment]
Apart from the one comment, LGTM. You can remove the code in separate commit (that can be appended to this pull request).

Other than that I'm wondering if this should go into M5. I'd feel better if we merged that after M5 because this change has a potential of breaking stuff in unexpected way (as usually with specialization).

File: src/compiler/scala/tools/nsc/transform/UnCurry.scala (hunk 1 above)
Comment (axel22) [inline src/compiler/scala/tools/nsc/transform/UnCurry.scala]: Good point.

**Comment by axel22** [comment]
I also have a tendency to have this merged after M5.

**Comment by axel22** [comment]
Not sured why strap failed, as it didn't fail in the manual checkin:

https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/203/console

**Comment by gkossakowski** [comment]
There it was building `6aefae4cbf`, here it was building `4cc4773`. Not sure what's the relationship between those commits.

**Comment by axel22** [comment]
Strange.

The only difference is the name of the test files, pretty much:

https://github.com/axel22/scala-github/commit/4cc4773fa4a086a80a590c497a01c7d695666d44

https://github.com/axel22/scala-github/commit/6aefae4cbfdf1f791b2c77bd5b97685bdaeaae03

**Comment by gkossakowski** [comment]
If you compare logs you'll notice that the successful build was doing more cleaning, this is the missing part from failed build:

```
Removing build/
Removing dists/
Removing jenkins-artifacts/
Removing jenkins-scripts/
Removing lib/ant/ant-contrib.jar
Removing lib/ant/ant-dotnet-1.0.jar
Removing lib/ant/ant.jar
Removing lib/ant/maven-ant-tasks-2.1.1.jar
Removing lib/ant/vizant.jar
Removing lib/extra/
Removing lib/fjbg.jar
Removing lib/forkjoin.jar
Removing lib/jline.jar
Removing lib/msil.jar
Removing lib/scala-compiler.jar
Removing lib/scala-library-src.jar
Removing lib/scala-library.jar
Removing lib/scala-reflect.jar
Removing test/files/codelib/code.jar
Removing test/files/lib/annotations.jar
Removing test/files/lib/enums.jar
Removing test/files/lib/genericNest.jar
Removing test/files/lib/methvsfield.jar
Removing test/files/lib/nest.jar
Removing test/files/lib/scalacheck.jar
Removing test/files/speclib/instrumented.jar
```

**Comment by axel22** [comment]
Yes, I saw that. I imagine there is a difference in `msil.jar`s, but I am not sure how to make the bot build do the cleaning.

**Comment by axel22** [comment]
That doesn't seem to be the reason, though, because with this commit the checkin manual also fails..

https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/204/consoleFull

**Comment by gkossakowski** [comment]
Then the only thing I can think of is that it's not you who introduced stability problem but one of the parent commit your commit is based on. I suggest rebasing on top of the one the that works (e.g. 026a70d55591c3b5ee157e22998b62168afee686, see https://scala-webapps.epfl.ch/jenkins/job/scala-nightly-main/1486/) and trying again.

**Comment by axel22** [comment]
A new rebased pull request has been issued.

**Pull request closed without merging**